util/progress/progressui/display.go

```diff
@@ -71,51 +81,146 @@ func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch
+			d.display(false)
+			ticker.Reset(tickerTimeout)
+		}
+	}
+}
+
+type consoleDisplay struct {
+	t             *trace
+	disp          *display
+	width, height int
+}
+
+func NewConsoleDisplay(c console.Console) SolveStatusDisplay {
+	return &consoleDisplay{
+		t:    newTrace(c, true),
+		disp: &display{c: c},
+	}
+}
+
+func (d *consoleDisplay) init(dsso *displaySolveStatusOpts) {
+	d.disp.phase = dsso.phase
+	d.disp.desc = dsso.consoleDesc
+	if d.disp.phase == "" {
+		d.disp.phase = "Building"
+	}
+	d.width, d.height = d.disp.getSize()
+}
+
+func (d *consoleDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, d.width)
+}
+
+func (d *consoleDisplay) display(done bool) {
+	d.width, d.height = d.disp.getSize()
+	d.disp.print(d.t.displayInfo(), d.width, d.height, done)
+	if done {
+		d.t.printErrorLogs(d.t.w)
+	}
+}
+
+func (d *consoleDisplay) warnings() []client.VertexWarning {
+	return d.t.warnings()
+}
+
+type textDisplay struct {
+	t       *trace
+	printer *textMux
+}
+
+func newTextDisplay(w io.Writer) SolveStatusDisplay {
+	return &textDisplay{
+		t:       newTrace(w, false),
+		printer: &textMux{w: w, printer: (*textPrinter)(nil)},
+	}
+}
+
+func newJsonDisplay(w io.Writer) SolveStatusDisplay {
+	return &textDisplay{
+		t:       newTrace(w, false),
+		printer: &textMux{w: w, printer: &jsonPrinter{}},
+	}
+}
+
+func (d *textDisplay) init(dsso *displaySolveStatusOpts) {
+	d.printer.desc = dsso.textDesc
+}
+
+func (d *textDisplay) update(ss *client.SolveStatus) {
+	d.t.update(ss, 80)
+}
+
+func (d *textDisplay) display(done bool) {
+	d.printer.print(d.t)
+	if done {
+		d.t.printErrorLogs(d.t.w)
+	}
+}
+
+func (d *textDisplay) warnings() []client.VertexWarning {
+	return d.t.warnings()
+}
+
+type DisplayMode string
+
+const (
+	AutoMode  DisplayMode = "auto"
+	TtyMode               = "tty"
+	PlainMode             = "plain"
+	JsonMode              = "json"
+)
+
+func NewDisplay(out console.File, mode DisplayMode) (SolveStatusDisplay, error) {
+	switch mode {
+	case AutoMode, TtyMode, "":
+		if c, err := console.ConsoleFromFile(out); err == nil {
+			return NewConsoleDisplay(c), nil
 		} else {
-			if done || displayLimiter.Allow() {
-				printer.print(t)
-				if done {
-					t.printErrorLogs(w)
-					return t.warnings(), nil
-				}
-				ticker.Stop()
-				ticker = time.NewTicker(tickerTimeout)
+			if mode == "tty" {
+				return nil, errors.Wrap(err, "failed to get console")
 			}
 		}
+		fallthrough
+	case PlainMode:
+		return newTextDisplay(out), nil
+	case JsonMode:
+		return newJsonDisplay(out), nil
+	default:
+		return nil, errors.Errorf("invalid progress mode %s", mode)
+	}
+}
+
+func DisplaySolveStatus(ctx context.Context, c console.Console, w io.Writer, ch chan *client.SolveStatus, opts ...DisplaySolveStatusOpt) ([]client.VertexWarning, error) {
+	var d SolveStatusDisplay
+	if c != nil {
+		d = NewConsoleDisplay(c)
+	} else {
+		d = newTextDisplay(w)
 	}
+	return UpdateDisplay(ctx, d, ch, opts...)
 }
```

Comment:
We can modify this if we want, but I think it will probably be better to keep this method around with its current signature until the places that use this have been updated. We can add a deprecation notice to it to ease that transition if it helps.

The new method is intended to be using `NewDisplay` to create the display and then `UpdateDisplay` to feed the display with the status updates.